error.go

```diff
@@ -20,295 +20,233 @@
-	})
-}
+	// ErrInternal is the class of error used for non HTTPError.
+	ErrInternal = &ErrorClass{errInternalID, "internal error", 500}
+)
 
-// Error builds an error message from the typed error details.
-func (t *TypedError) Error() string {
-	IncrCounter([]string{"goa", "error", strconv.Itoa(int(t.ID))}, 1.0)
-	js, err := json.Marshal(t)
-	if err != nil {
-		return `{"id":0,"title":"generic","msg":"failed to serialize error"}`
+type (
+	// HTTPError describes an error that can be returned in a response.
+	HTTPError struct {
+		// ErrorClass defines static properties for all error instances that use it.
+		*ErrorClass
+		// Err describes the specific error occurrence.
+		Err string `json:"err" xml:"err"`
 	}
-	return string(js)
-}
 
-// Error summarizes all the underlying error messages in one JSON array.
-func (m MultiError) Error() string {
-	var buffer bytes.Buffer
-	buffer.WriteString("[")
-	for i, err := range m {
-		txt := err.Error()
-		if _, ok := err.(*TypedError); !ok {
-			b, err := json.Marshal(txt)
-			txt = `{"id":0,"title":"generic","msg":`
-			if err != nil {
-				txt += `"unknown error"}`
-			} else {
-				txt += string(b) + "}"
-			}
-		}
-		buffer.WriteString(txt)
-		if i < len(m)-1 {
-			buffer.WriteString(",")
-		}
+	// ErrorClass contains information sent together with the error message in responses.
+	ErrorClass struct {
+		// ID identifies the class of errors for client programs.
+		ID string `json:"id" xml:"id"`
+		// Title is a human friendly title for the class of errors.
+		Title string `json:"title" xml:"title"`
+		// Status is the HTTP status code used by responses that cary the error.
+		Status int `json:"-" xml:"-"`
 	}
-	buffer.WriteString("]")
 
-	// you can blame rsc for that: https://code.google.com/p/go/issues/detail?id=8592#c3
-	txt := buffer.String()
-	txt = strings.Replace(txt, "\\u0026", "&", -1)
-	txt = strings.Replace(txt, "\\u003c", "<", -1)
-	txt = strings.Replace(txt, "\\u003e", ">", -1)
+	// MultiError is an error composed of potentially multiple errors.
+	MultiError []error
+)
 
-	return txt
+// NewErrorClass creates a new error class.
+// It is the responsability of the client to guarantee uniqueness of id.
+func NewErrorClass(id, title string, status int) *ErrorClass {
+	return &ErrorClass{ID: id, Title: title, Status: status}
 }
 
-// NewBadRequestError wraps the given error into a BadRequestError.
-func NewBadRequestError(err error) *BadRequestError {
-	return &BadRequestError{Actual: err}
+// Error wraps an error into a HTTP error of the given class.
+func (class *ErrorClass) Error(err error) *HTTPError {
+	return &HTTPError{ErrorClass: class, Err: err.Error()}
 }
 
-// Error implements error.
-func (b *BadRequestError) Error() string {
-	return b.Actual.Error()
+// Errorf builds an HTTP error given a format and values.
+func (class *ErrorClass) Errorf(fm string, v ...interface{}) *HTTPError {
+	msg := fmt.Sprintf(fm, v...)
+	return &HTTPError{ErrorClass: class, Err: msg}
 }
```

Comment:
those two (`Error` and `Errorf`) are superb ! I love it... simple and extensible..